hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,857 @@
+    ///
+    /// `payload` is just a way for the transfer to own resources
+    /// while the transfer is ongoing. For instance, a SERCOM instance could be
+    /// moved into the transfer to prevent data races until the transfer is
+    /// complete. Calling [`wait`](DmaTransfer::wait) will release the
+    /// payload for reuse.
+    ///
+    /// Here, we use array references instead of slices to *statically* ensure
+    /// they are the same length. If the two buffers weren't the same length,
+    /// we could run into undefined behaviour, especially if the destination
+    /// buffer is shorter than the source buffer. In that specific case, the
+    /// DMAC would end up overrunning the destination buffer and would write
+    /// into a memory section we don't own.
+    ///
+    /// # Safety
+    ///
+    /// This function is `unsafe`. You must either manually guarantee that both
+    /// `source` and `dest` are `'static`, or that the returned
+    /// `UnsafeTransfer` WILL be stopped (by calling `stop`) or waited upon
+    /// (by calling `wait`) before being dropped in every possible case.
+    #[inline(always)]
+    pub unsafe fn inc_src_inc_dest_unchecked(
+        chan: Channel<Ready, ID>,
+        source: &'a mut [B; N],
+        dest: &'a mut [B; N],
+        circular_xfer: bool,
+        payload: P,
+    ) -> Self {
+        let length = source.len() as u16;
+        let config = TransferConfiguration {
+            source,
+            dest,
+            length,
+            _t: PhantomData,
+        };
+        Self::setup(chan, config, circular_xfer, payload)
+    }
+}
+
+/// These methods offer a transfer from an fixed source to a fixed destination
+impl<B: 'static + DmaBeat, P, const ID: u8>
+    DmaTransfer<B, P, &'static mut B, &'static mut B, Ready, ID>
+{
+    /// Fixed source to fixed destination.
+    ///
+    /// If `circular_xfer == false`, the transfer will run once
+    /// and stop. If `circular_xfer == true`, then the transfer will be set as
+    /// circular, meaning it will restart a new, identical block transfer
+    /// when the current block transfer is complete.
+    ///
+    /// `payload` is just a way for the transfer to own resources
+    /// while the transfer is ongoing. For instance, a SERCOM instance could be
+    /// moved into the transfer to prevent data races until the transfer is
+    /// complete. Calling [`wait`](DmaTransfer::wait) will release the
+    /// payload for reuse.
+    #[inline(always)]
+    pub fn fixed_src_fixed_dest(
+        chan: Channel<Ready, ID>,
+        source: &'static mut B,
+        dest: &'static mut B,
+        circular_xfer: bool,
+        payload: P,
+    ) -> Self {
+        let length = 1;
+        let config = TransferConfiguration {
+            source,
+            dest,
+            length,
+            _t: PhantomData,
+        };
+        DmaTransfer {
+            xfer: UnsafeTransfer::setup(chan, config, circular_xfer, payload),
+        }
```

Comment:
In every wrapper function, you calculate a length, create a `TransferConfiguration` and then a `Transfer` of some type. I think it should be possible to do that all generically. This may not be the best idea, but it's one that comes to mind immediately. You could create a trait that is implemented on `(Src, Dst)` pairs. It would have a `length` function that would calculate the transfer length based on the tuple types.